addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjection.java

```diff
@@ -0,0 +1,705 @@
+                                "Caught "
+                                        + ex.getClass().getName()
+                                        + " "
+                                        + ex.getMessage()
+                                        + " when "
+                                        + URI_EX_LOG);
+                    }
+                    return;
+                }
+            }
+        }
+        // injection attack to $Where and $mapReduce cluases
+        // The $where clause executes associated js function one time for each tuple --> sleep time
+        // = interval * nTuples
+        if (doTimedScan) {
+            if (isStop()) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug(STOP_LOG);
+                }
+                return;
+            }
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("\nStarting with the javascript code injection package of attack rules:");
+            }
+            int aveRtt = getAveRtts(rtt),
+                    index = 0,
+                    timeShort = SLEEP_SHORT_TIME,
+                    timeLong = SLEEP_LONG_TIME;
+            String phase = null;
+            boolean hadTimeout = false;
+            String sleepValueToInj;
+            while (index < SLEEP_INJECTION.length) {
+                if (isStop()) {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug(STOP_LOG);
+                    }
+                    return;
+                }
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("\nTrying  with the value: " + SLEEP_INJECTION[index][0]);
+                }
+                try {
+                    msgInjAttack = getNewMsg();
+                    sleepValueToInj =
+                            SLEEP_INJECTION[index][0].replaceFirst(
+                                    INSERT_SHORT_TIME, Integer.toString(timeShort));
+                    phase = INSERT_SHORT_TIME;
+                    setParameter(msgInjAttack, param, sleepValueToInj);
+                    sendAndReceive(msgInjAttack, false);
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug(
+                                "\nTrying for a longer time with the value: "
+                                        + SLEEP_INJECTION[index][1]);
+                    }
+                    if (msgInjAttack.getTimeElapsedMillis() >= aveRtt + DELTA_TIME) {
+                        phase = INSERT_LONG_TIME;
+                        sleepValueToInj =
+                                SLEEP_INJECTION[index][1].replaceFirst(
+                                        INSERT_LONG_TIME, Integer.toString(timeLong));
+                        msgInjAttack = getNewMsg();
+                        setParameter(msgInjAttack, param, sleepValueToInj);
+                        sendAndReceive(msgInjAttack, false);
+                        if (msgInjAttack.getTimeElapsedMillis() >= aveRtt + DELTA_TIME) {
```

Comment:
Hello @LuigiCasciaro. Is there any reason to do the same comparison as in line 464? If the vulnerability exists the delay time will be bigger because you are sending a bigger amount of time (unless there is something restricting the maximum time to response). I think that the comparison could be something as ``` msgInjAttack.getTimeElapsedMillis() >= aveRtt + DELTA_TIME + (SLEEP_LONG_TIME - SLEEP_SHORT_TIME) ```.

Reply:
Hello @DiogoMrSilva. The reason of this implementation is that the mongo sleep is performed once for each tuple so any speculation about time turns out to be ineffective, and in fact the true meaning more than of `DELTA_TIME` (name that made sense in old versions) is of `THRESHOLD`.

That said, surely using the same variable is incorrect (thank you very much for reporting it) and it would be more correct to distinguish the two cases: `aveRtt +  THRESHOLD_SHORT_TIME` and `aveRtt +  THRESHOLD_LONG_TIME`.

I don't understand why your proposal: `(SLEEP_LONG_TIME - SLEEP_SHORT_TIME)`, could you explain it to me?

**FYI**:
The main scenario is that a timeout occurs since a relatively small collection would be enough to exceed the maximum WT (30 or 40 seconds I don't remember exactly). 

So the strategy implemented is the following: when a timeout occurs for the first time, depending on whether you have tried for the SHORT or LONG time, the injected time is divided by 3 or 10 respectively; 3 because we are relatively close since in this case the SHORT was successful; 10 instead is quite random but the aim is to reduce a lot since the attempt with `SHORT_TIME` has already timed out.

Reply:
Thx for the information @LuigiCasciaro, now I understand better what you did. I did the wrong assumption that the sleep would just be executed once. 

The idea behind the using  `aveRtt + DELTA_TIME` and after `aveRtt + DELTA_TIME + (SLEEP_LONG_TIME - SLEEP_SHORT_TIME)` is the same as using `aveRtt + SLEEP_SHORT_TIME` and after `aveRtt + SLEEP_LONG_TIME` it increases the time expected by `SLEEP_LONG_TIME - SLEEP_SHORT_TIME`.